lib/logsource.c

```diff
@@ -311,10 +310,11 @@ _reclaim_window_instead_of_rebalance(LogSource *self)
       self->full_window_size -= total_reclaim;
       stats_counter_sub(self->stat_full_window, total_reclaim);
       dynamic_window_release(&self->dynamic_window, total_reclaim);
-      if ((gssize)atomic_gssize_get(&self->window_size_to_be_reclaimed) > 0)
-        {
-          reclaim_in_progress = TRUE;
-        }
+      gssize to_be_reclaimed = (gssize)atomic_gssize_get(&self->window_size_to_be_reclaimed);
+      reclaim_in_progress = to_be_reclaimed > 0;
+      //to avoid underflow, we need to set a value <= 0
+      if (to_be_reclaimed < 0)
+        atomic_gssize_set(&self->window_size_to_be_reclaimed, 0);
```

Comment:
This only reduces the possibility of an overflow, because `_take_reclaimed_window()` (the function containing the subtraction) is called for every ACK even when reclaim is not in progress, while this sets `window_size_to_be_reclaimed` to 0 only when `pending_reclaimed` is nonzero.

I think `atomic_gssize_set(&self->window_size_to_be_reclaimed, 0);` should be moved out of the `if (total_reclaim > 0)` branch.